File: csharp/Minesweeper.Tests/UnitTests/GridTests.cs
using MineSweeper.Engine;

namespace MineSweeper.Tests.UnitTests
{
    [TestFixture]
    public class GridTests
    {
        [Test]
        public void GridInitialization_CorrectSizeAndCells()
        {
            int size = 4;
            var grid = new Grid(size);
            Assert.Multiple(() =>
            {
                Assert.That(grid.Size, Is.EqualTo(size));
                Assert.That(grid.Cells, Is.Not.Null);
            });
        }

        [Test]
        public void GridInitialization_AllCellsNotMinedAndNotRevealed()
        {
            var grid = new Grid(4); // Example size of 4x4
            bool allCellsCorrect = true;
            for (int i = 0; i < grid.Size; i++)
            {
                for (int j = 0; j < grid.Size; j++)
                {
                    if (grid.Cells[i, j].IsMine || grid.Cells[i, j].IsRevealed)
                    {
                        allCellsCorrect = false;
                        break;
                    }
                }
                if (!allCellsCorrect) break;
            }
            Assert.That(allCellsCorrect, "All cells should be initialized without mines and not revealed.");
        }


        [Test]
        public void PlaceMines_CorrectNumberOfMinesPlaced()
        {
            var grid = new Grid(5); // 5x5 grid
            int numberOfMines = 5;
            grid.PlaceMines(numberOfMines);
            int mineCount = 0;

            foreach (var cell in grid.Cells)
            {
                if (cell.IsMine)
                    mineCount++;
            }

            Assert.That(mineCount, Is.EqualTo(numberOfMines), "The number of mines placed should match the specified amount.");
        }


        [Test]
        public void UncoverCell_NoAdjacentMines_UncoverSurroundingCells()
        {
            var grid = new Grid(3);
            grid.PlaceMinesManually([new(2, 2)]);
            grid.UncoverCell(0, 0); // Uncover the top-left corner, which should have no adjacent mines

            // Check surrounding cells are uncovered
            Assert.That(grid.Cells[0, 0].IsRevealed && grid.Cells[0, 1].IsRevealed && grid.Cells[1, 0].IsRevealed,
                "Surrounding cells should be automatically uncovered if no adjacent mines.");
        }
        
        [Test]
        public void InitializeCells_ShouldCreateCellsOfCorrectSize()
        {
            // Arrange
            int size = 5;
            var grid = new Grid(size);

            Assert.Multiple(() =>
            {
                // Act & Assert
                Assert.That(grid.Cells.GetLength(0), Is.EqualTo(size));
                Assert.That(grid.Cells.GetLength(1), Is.EqualTo(size));
            });
        }

        [Test]
        public void PlaceMines_ShouldPlaceCorrectNumberOfMines()
        {
            // Arrange
            int size = 5;
            int numberOfMines = 5;
            var grid = new Grid(size);

            // Act
            grid.PlaceMines(numberOfMines);

            int mineCount = 0;
            for (int i = 0; i < size; i++)
            {
                for (int j = 0; j < size; j++)
                {
                    if (grid.Cells[i, j].IsMine)
                    {
                        mineCount++;
                    }
                }
            }

            // Assert
            Assert.That(mineCount, Is.EqualTo(numberOfMines));
        }

        [Test]
        public void PlaceMinesManually_ShouldPlaceMinesAtSpecifiedPositions()
        {
            // Arrange
            int size = 5;
            var minePositions = new List<Point> { new(0, 0), new(1, 1), new(2, 2) };
            var grid = new Grid(size);

            // Act
            grid.PlaceMinesManually(minePositions);

            // Assert
            foreach (var position in minePositions)
            {
                Assert.That(grid.Cells[position.Row, position.Col].IsMine);
            }
        }

        [Test]
        public void UncoverCell_ShouldUncoverCellAndReturnCorrectCell()
        {
            // Arrange
            int size = 5;
            var grid = new Grid(size);
            int row = 2;
            int col = 2;

            // Act
            var uncoveredCell = grid.UncoverCell(row, col);

            Assert.Multiple(() =>
            {
                // Assert
                Assert.That(uncoveredCell.IsRevealed);
                Assert.That(uncoveredCell.AdjacentMines, Is.EqualTo(0));
            });
        }

        [Test]
        public void UncoverAdjacentCells_ShouldUncoverAdjacentCells()
        {
            // Arrange
            int size = 5;
            var grid = new Grid(size);
            int row = 2;
            int col = 2;

            // Act
            grid.UncoverAdjacentCells(row, col);

            // Assert
            for (int i = row - 1; i <= row + 1; i++)
            {
                for (int j = col - 1; j <= col + 1; j++)
                {
                    if (i >= 0 && i < size && j >= 0 && j < size)
                    {
                        Assert.That(grid.Cells[i, j].IsRevealed);
                    }
                }
            }
        }

        [Test]
        public void CountAdjacentMines_ShouldReturnCorrectCount()
        {
            // Arrange
            int size = 5;
            var grid = new Grid(size);
            int row = 2;
            int col = 2;

            // Act
            int adjacentMines = grid.CountAdjacentMines(row, col);

            // Assert
            Assert.That(adjacentMines, Is.EqualTo(0));
        }

        [Test]
        public void UncoverAdjacentCells_PointOverload_ShouldUncoverAdjacentCells()
        {
            // Arrange
            int size = 5;
            var grid = new Grid(size);
            var point = new Point(2, 2);

            // Act
            grid.UncoverAdjacentCells(point);

            // Assert
            for (int i = point.Row - 1; i <= point.Row + 1; i++)
            {
                for (int j = point.Col - 1; j <= point.Col + 1; j++)
                {
                    if (i >= 0 && i < size && j >= 0 && j < size)
                    {
                        Assert.That(grid.Cells[i, j].IsRevealed);
                    }
                }
            }
        }

        [Test]
        public void UncoverCell_PointOverload_ShouldUncoverCellAndReturnCorrectCell()
        {
            // Arrange
            int size = 5;
            var grid = new Grid(size);
            var point = new Point(2, 2);

            // Act
            var uncoveredCell = grid.UncoverCell(point);

            Assert.Multiple(() =>
            {
                // Assert
                Assert.That(uncoveredCell.IsRevealed);
                Assert.That(uncoveredCell.AdjacentMines, Is.EqualTo(0));
            });
        }

        [Test]
        public void CountAdjacentMines_PointOverload_ShouldReturnCorrectCount()
        {
            // Arrange
            int size = 5;
            var grid = new Grid(size);
            var point = new Point(2, 2);

            // Act
            int adjacentMines = grid.CountAdjacentMines(point);

            // Assert
            Assert.That(adjacentMines, Is.EqualTo(0));
        }

        [Test]
        public void GetCell_ShouldReturnCorrectCell()
        {
            // Arrange
            int size = 5;
            var grid = new Grid(size);
            int row = 2;
            int col = 2;

            // Act
            var cell = grid.GetCell(row, col);

            // Assert
            Assert.That(cell, Is.EqualTo(grid.Cells[row, col]));
        }

        [Test]
        public void GetCell_PointOverload_ShouldReturnCorrectCell()
        {
            // Arrange
            int size = 5;
            var grid = new Grid(size);
            var point = new Point(2, 2);

            // Act
            var cell = grid.GetCell(point);

            // Assert
            Assert.That(cell, Is.EqualTo(grid.Cells[point.Row, point.Col]));
        }
    }

}

File: kotlin/src/test/kotlin/unittests/GridTests.kt
package unittests

import org.minesweeper.engine.Grid
import org.minesweeper.engine.Point

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

internal class GridTests {

    @Test
    fun `Grid initialization - Correct size and cells`() {
        val size = 4
        val grid = Grid(size)
        assertAll("Checking grid initialization",
            { assertEquals(size, grid.size) },
            { assertNotNull(grid.cells) }
        )
    }

    @Test
    fun `Grid initialization - All cells not mined and not revealed`() {
        val grid = Grid(4)  // Example size of 4x4
        val allCellsCorrect = grid.cells.all { row -> row.all { cell -> !cell.isMine && !cell.isRevealed } }
        assertTrue(allCellsCorrect, "All cells should be initialized without mines and not revealed.")
    }

    @Test
    fun `Place mines - Correct number of mines placed`() {
        val grid = Grid(5)  // 5x5 grid
        val numberOfMines = 5
        grid.placeMines(numberOfMines)
        val mineCount = grid.cells.flatten().count { it.isMine }
        assertEquals(numberOfMines, mineCount, "The number of mines placed should match the specified amount.")
    }

    @Test
    fun `Uncover cell - No adjacent mines, uncover surrounding cells`() {
        val grid = Grid(3)
        grid.placeMinesManually(listOf(Point(2, 2)))  // Only mine at the bottom-right corner
        grid.uncoverCell(0, 0)  // Uncover the top-left corner, which should have no adjacent mines

        // Check surrounding cells are uncovered
        assertTrue(grid.cells[0][0].isRevealed && grid.cells[0][1].isRevealed && grid.cells[1][0].isRevealed,
            "Surrounding cells should be automatically uncovered if no adjacent mines.")
    }

    @Test
    fun `Initialize cells - Should create cells of correct size`() {
        val size = 5
        val grid = Grid(size)
        assertAll("Checking cell array dimensions",
            { assertEquals(size, grid.cells.size) },
            { assertTrue(grid.cells.all { it.size == size }) }
        )
    }

    @Test
    fun `Place mines manually - Should place mines at specified positions`() {
        val size = 5
        val minePositions = listOf(Point(0, 0), Point(1, 1), Point(2, 2))
        val grid = Grid(size)
        grid.placeMinesManually(minePositions)

        minePositions.forEach { position ->
            assertTrue(grid.cells[position.row][position.col].isMine, "Mine should be placed at (${position.row}, ${position.col})")
        }
    }

    @Test
    fun `Uncover cell - Should uncover cell and return correct cell`() {
        val size = 5
        val grid = Grid(size)
        val row = 2
        val col = 2

        val uncoveredCell = grid.uncoverCell(row, col)
        assertAll("Checking uncovered cell properties",
            { assertTrue(uncoveredCell.isRevealed) },
            { assertEquals(0, uncoveredCell.adjacentMines) }
        )
    }

    @Test
    fun `Uncover adjacent cells - Should uncover adjacent cells`() {
        val size = 5
        val grid = Grid(size)
        val row = 2
        val col = 2

        grid.uncoverAdjacentCells(row, col)
        for (i in row - 1..row + 1) {
            for (j in col - 1..col + 1) {
                if (i in 0 until size && j in 0 until size) {
                    assertTrue(grid.cells[i][j].isRevealed, "Cell at ($i, $j) should be uncovered")
                }
            }
        }
    }

    @Test
    fun `Count adjacent mines - Should return correct count`() {
        val size = 5
        val grid = Grid(size)
        val row = 2
        val col = 2

        val adjacentMines = grid.countAdjacentMines(row, col)
        assertEquals(0, adjacentMines, "Should count zero adjacent mines for cell at (2, 2)")
    }

    @Test
    fun `Uncover adjacent cells - Point overload - Should uncover adjacent cells`() {
        val size = 5
        val grid = Grid(size)
        val point = Point(2, 2)

        grid.uncoverAdjacentCells(point)
        for (i in point.row - 1..point.row + 1) {
            for (j in point.col - 1..point.col + 1) {
                if (i in 0 until size && j in 0 until size) {
                    assertTrue(grid.cells[i][j].isRevealed, "Cell at ($i, $j) should be uncovered")
                }
            }
        }
    }

    @Test
    fun `Uncover cell - Point overload - Should uncover cell and return correct cell`() {
        val size = 5
        val grid = Grid(size)
        val point = Point(2, 2)

        val uncoveredCell = grid.uncoverCell(point)
        assertAll("Checking uncovered cell properties using Point overload",
            { assertTrue(uncoveredCell.isRevealed) },
            { assertEquals(0, uncoveredCell.adjacentMines) }
        )
    }

    @Test
    fun `Count adjacent mines - Point overload - Should return correct count`() {
        val size = 5
        val grid = Grid(size)
        val point = Point(2, 2)

        val adjacentMines = grid.countAdjacentMines(point)
        assertEquals(0, adjacentMines, "Should count zero adjacent mines for cell at (2, 2) using Point overload")
    }

    @Test
    fun `Get cell - Should return correct cell`() {
        val size = 5
        val grid = Grid(size)
        val row = 2
        val col = 2

        val cell = grid.getCell(row, col)
        assertEquals(grid.cells[row][col], cell, "Should return the correct cell at (2, 2)")
    }

    @Test
    fun `Get cell - Point overload - Should return correct cell`() {
        val size = 5
        val grid = Grid(size)
        val point = Point(2, 2)

        val cell = grid.getCell(point)
        assertEquals(grid.cells[point.row][point.col], cell, "Should return the correct cell at (2, 2) using Point overload")
    }
}
